src/Backend.idr

```diff
@@ -12,86 +12,49 @@ import Text.PrettyPrint.WL
-getUsedVars : Env n -> (td: TDef n) -> Env (length (getUsedIndices td))
-getUsedVars e td = map (flip index e) (fromList $ getUsedIndices td)
+||| Interface for interpreting type definitions as ASTs.
+||| @def  the type representing type definitions.
+||| @type the type representing type signatures.
+||| @n    the number of variables the interpretor supports in a type definition. (Should always be either `n` or `0`.)
+interface ASTGen def type (n : Nat) | def where
+  ||| Given a `TDef`, generate its corresponding type signature.
+  msgType        : TNamed n -> type
 
-||| Interface for codegens. `lang` is a type representing (the syntactic structure of)
-||| a type declaration in the target language.
-interface Backend lang where
-  ||| Given a `TDef` and a matching environment, generate a list of type definitions
-  ||| in the target language.
-  generateTyDefs : Env n -> TDef n -> List lang
+  ||| Generate definitions for a `TNamed` and all its helper definitions.
+  generateTyDefs : TNamed n -> List def
 
-  ||| Given a type definition in the target language, generate its code.
-  generateCode : lang -> Doc
+||| Interface for code generators that can generate code for type definitions and
+||| type signatures independently of each other.
+||| @def  the type representing type definitions.
+||| @type the type representing type signatures.
+interface CodegenIndep def type | def where
+  ||| Generate source code for a type signature.
+  typeSource : type -> Doc
 
-  ||| Generate a new environment with n free names.
-  freshEnv : (n: Nat) -> Env n
+  ||| Generate source code for a type definition.
+  defSource  : def -> Doc
 
 ||| Use the given backend to generate code for a type definition and all its dependencies.
-generate : (lang: Type) -> Backend lang => {n: Nat} -> TDef n -> Doc
-generate lang {n} td = vsep2 . map (generateCode) . generateTyDefs {lang} (freshEnv {lang} n) $ td
-
-||| Interface for codegens which distinguish between the message type itself and
-||| its helper definitions. Currently doesn't support typedefs with free variables. 
-interface NewBackend def type | def where
-  ||| Given a `TDef`, generate its corresponding type signature.
-  msgType  : TDef 0 -> type
+generateDefs : (def: Type) -> ASTGen def type n => CodegenIndep def type => TNamed n -> Doc
+generateDefs def tn = vsep2 $ map defSource (generateTyDefs {def} tn)
 
-  ||| Given a `TDef`, generate a list of all helper definitions required by it.
-  typedefs : TDef 0 -> List def
+||| Use the given backend to generate code for a type signature.
+generateType : (def: Type) -> ASTGen def type n => CodegenIndep def type => TNamed n -> Doc
+generateType def tn = typeSource {def} (msgType {def} tn)
 
-  ||| Given a type signature and a list of helper definitions which it uses,
-  ||| generate source code.
-  source   : type -> List def -> Doc
+||| Interface for code generators that need to generate code for type definitions and
+||| type signatures at the same time.
```

Comment:
Similarly, it might be nice to list some examples also here, e.g. from the PR description 
"JSON".